### PR TITLE
C++: Remove FPs from cpp/uninitialized-local when encountered extraction errors 

### PR DESCRIPTION
**Pull request by calumgrant**
Fixes false-positives in standalone extraction mode, whereby local variables may be initialized using function calls to unrecognised functions. These manifest as `ErrorExpr`, so we can exclude these from the results.

The DCA results are interesting. We lose one FP on abseil (this is good), but also a bunch of results on the samate project, which are due to an `ErrorExpr` creeping into one of the C functions, which are all called `goodB2G` or `bad`, so seem to identify as the same function. However, the root cause of the `ErrorExpr` is unclear as there's nothing obvious in the extractor logs.

Questions:
1. Are we ok with the DCA results? Should we merge this as-is?
2. Change-note required? (Will fail PR-check)

#### All query authors

- [ ] ~~A change note is added if necessary. See [the documentation](https://github.com/github/codeql/blob/main/docs/change-notes.md) in this repository.~~
- [ ] ~~All new queries have appropriate `.qhelp`. See [the documentation](https://github.com/github/codeql/blob/main/docs/query-help-style-guide.md) in this repository.~~
- [x] QL tests are added if necessary. See [Testing custom queries](https://docs.github.com/en/code-security/codeql-cli/using-the-advanced-functionality-of-the-codeql-cli/testing-custom-queries) in the GitHub documentation.
- [ ] ~~New and changed queries have correct query metadata. See [the documentation](https://github.com/github/codeql/blob/main/docs/query-metadata-style-guide.md) in this repository.~~

#### Internal query authors only

- [ ] ~~Autofixes generated based on these changes are valid, only needed if this PR makes significant changes to `.ql`, `.qll`, or `.qhelp` files. See [the documentation](https://github.com/github/codeql-team/blob/main/docs/best-practices/validating-autofix-for-query-changes.md) (internal access required).~~
- [ ] Changes are validated [at scale](https://github.com/github/codeql-dca/) (internal access required).
- [ ] ~~Adding a new query? Consider also [adding the query to autofix](https://github.com/github/codeml-autofix/blob/main/docs/updating-query-support.md#adding-a-new-query-to-the-query-suite).~~

### CODE REVIEW

File: cpp/ql/test/query-tests/Security/CWE/CWE-457/semmle/Errors/UninitializedLocal.qlref [hunk 1]
```diff
@@ -0,0 +1 @@
+Likely Bugs/Memory Management/UninitializedLocal.ql
```
Comment (jketema) [inline cpp/ql/test/query-tests/Security/CWE/CWE-457/semmle/Errors/UninitializedLocal.qlref:1]: The tests for this query are located in `cpp/test/query-tests/Security/CWE/CWE-457/semmle/tests`. We should probably add the test case there.

Comment (calumgrant) [reply cpp/ql/test/query-tests/Security/CWE/CWE-457/semmle/Errors/UninitializedLocal.qlref:1]: Fine by me.

**Comment by jketema** [comment]
> The DCA results are interesting. We lose one FP on abseil (this is good), but also a bunch of results on the samate project, which are due to an `ErrorExpr` creeping into one of the C functions, which are all called `goodB2G` or `bad`, so seem to identify as the same function. However, the root cause of the `ErrorExpr` is unclear as there's nothing obvious in the extractor logs.

This is for the same reason as: https://github.com/github/codeql/blob/d0831ebd5a493586e6581f13878a7ecce704b7d3/cpp/ql/lib/semmle/code/cpp/ir/implementation/raw/internal/TranslatedElement.qll#L151-L157

So we need to be a bit more careful here. We should probably define a predicate that captures this case and also use it in other queries that check for `ErrorExpr`.

**Comment by calumgrant** [comment]
> > The DCA results are interesting. We lose one FP on abseil (this is good), but also a bunch of results on the samate project, which are due to an `ErrorExpr` creeping into one of the C functions, which are all called `goodB2G` or `bad`, so seem to identify as the same function. However, the root cause of the `ErrorExpr` is unclear as there's nothing obvious in the extractor logs.
> 
> This is for the same reason as:
> 
> https://github.com/github/codeql/blob/d0831ebd5a493586e6581f13878a7ecce704b7d3/cpp/ql/lib/semmle/code/cpp/ir/implementation/raw/internal/TranslatedElement.qll#L151-L157
> 
> So we need to be a bit more careful here. We should probably define a predicate that captures this case and also use it in other queries that check for `ErrorExpr`.

That certainly matches what I've observed. Thanks for the pointer.

**Comment by jketema** [comment]
Note that there's something similar going on with the first argument of a deallocator call of a delete expressions. And I'm now left wondering whether we handle that correctly in the IR (will briefly investigate).

**Comment by calumgrant** [comment]
> Note that there's something similar going on with the first argument of a deallocator call of a delete expressions. And I'm now left wondering whether we handle that correctly in the IR (will briefly investigate).

This does leave me wondering why this expression is emitted at all. I can work around it for now, but a follow-up could be to replace this with something else or avoid it entirely.

I've not looked into the IR, but it does seem to me that `ignoreExprOnly(Expr expr)` should also consider `DeleteOrDeleteArrayExpr`.

**Comment by jketema** [comment]
> I've not looked into the IR, but it does seem to me that `ignoreExprOnly(Expr expr)` should also consider `DeleteOrDeleteArrayExpr`.

That only makes sense if the IR actually looks at that argument during IR generation, which I'm not sure of.

> This does leave me wondering why this expression is emitted at all. I can work around it for now, but a follow-up could be to replace this with something else or avoid it entirely.

The expression is an argument to a call, so something needs to be there. Looking briefly at the extractor code it didn't seem completely straightforward to replace this with what it really should be.

**Comment by jketema** [comment]
> I've not looked into the IR, but it does seem to me that ignoreExprOnly(Expr expr) should also consider DeleteOrDeleteArrayExpr.

I was mistaken. We only output the error expression in the case of new, and not delete. The extractor code is pretty difficult to follow here.

**Comment by calumgrant** [comment]
> > I've not looked into the IR, but it does seem to me that ignoreExprOnly(Expr expr) should also consider DeleteOrDeleteArrayExpr.
> 
> I was mistaken. We only output the error expression in the case of new, and not delete. The extractor code is pretty difficult to follow here.

I couldn't create a test case that generated the deallocator arg, so I decided to ignore it.

File: cpp/ql/lib/semmle/code/cpp/exprs/Expr.qll [hunk 2]
```diff
@@ -744,6 +744,13 @@ class ErrorExpr extends Expr, @errorexpr {
   override string toString() { result = "<error expr>" }
 
   override string getAPrimaryQlClass() { result = "ErrorExpr" }
+
+  /**
+   * Holds if this error expression is the first argument to a `new` allocation call.
+   */
+  predicate isFirstAllocatorCallArgument() {
+    this = any(NewOrNewArrayExpr new).getAllocatorCall().getArgument(0)
+  }
```
[the remark this answers is not in the document]
Comment (calumgrant) [reply cpp/ql/lib/semmle/code/cpp/exprs/Expr.qll:753]: I wasn't sure about this name. You suggested I should encapsulate the functionality from `TranslatedElement.qll`, but I'm not sure how helpful this really is. Could also rename it to `isExpected()`, but it's still just a workaround really. Happy to remove this predicate and inline it as it's not really adding much value.

**Comment by jketema** [comment]
> I couldn't create a test case that generated the deallocator arg, so I decided to ignore it.

Yup, because: "We only output the error expression in the case of new, and not delete."

File: cpp/ql/lib/semmle/code/cpp/exprs/Expr.qll (hunk 2 above)
[the remark this answers is not in the document]
Comment (jketema) [reply cpp/ql/lib/semmle/code/cpp/exprs/Expr.qll:753]: Sorry for the misunderstanding, I did not suggest encapsulating the functionality form `TranslatedElement.qll`. I only tried to suggest to do something similar. I don't think we need this predicate for that. It's something that can just be inlined in the `hasErrors` predicate you introduced.

File: cpp/ql/src/Likely Bugs/Memory Management/UninitializedLocal.ql [hunk 3]
```diff
@@ -89,5 +89,6 @@ from
 where
   conf.hasFlowPath(source, sink) and
   isSinkImpl(sink.getInstruction(), va) and
-  v = va.getTarget()
+  v = va.getTarget() and
+  not v.getFunction().hasErrors()
```
Comment (MathiasVP) [inline cpp/ql/src/Likely Bugs/Memory Management/UninitializedLocal.ql:93]: Might as well stick this constraint into `isSinkImpl` to speed up the recursion

File: cpp/ql/src/Best Practices/Unused Entities/UnusedLocals.ql [hunk 4]
```diff
@@ -57,5 +57,5 @@ where
   not declarationHasSideEffects(v) and
   not exists(AsmStmt s | f = s.getEnclosingFunction()) and
   not v.getAnAttribute().getName() = "unused" and
-  not any(ErrorExpr e).getEnclosingFunction() = f // unextracted expr may use `v`
+  not f.hasErrors()
```
Comment (jketema) [inline cpp/ql/src/Best Practices/Unused Entities/UnusedLocals.ql:60]: I think the comment should be retained here.

Comment (jketema) [reply cpp/ql/src/Best Practices/Unused Entities/UnusedLocals.ql:60]: Or at least something along those lines.

**Review by jketema — COMMENTED** [review]
LGTM. Just want to see some updated DCA results.

**Comment by calumgrant** [comment]
Will add change note.

**Review by jketema — APPROVED** [review]

**Pull request merged**